### PR TITLE
Default parameters of decision procedures

### PR DESCRIPTION
**Pull request by vhavlena**
This PR implements suitability checking for Nielsen decision procedure and underapproximation based on a concrete input instance. `try_nielsen` and `underapprox` are both set to true by default and they are applied on an instance if the instance is suitable for them (where it likely helps).

### CODE REVIEW

File: src/smt/theory_str_noodler/theory_str_noodler_final_check.cpp [hunk 1]
```diff
@@ -418,4 +418,35 @@ namespace smt::noodler {
...
+                
+                if(aut_ass.at(var)->size() <= 1) {
+                    continue;
+                }
+                if(aut_ass.is_co_finite(var, ln)) {
+                    continue;
+                }
+                if(aut_ass.is_singleton(var)) {
+                    continue;
+                }
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler_final_check.cpp:446]: Some reasoning why this condition would be good.

**Review by jurajsic — APPROVED** [review]

**Review by Adda0 — APPROVED** [review]
The changes look good to me. I second the suggestion made by Juraj.

File: src/smt/theory_str_noodler/theory_str_noodler_final_check.cpp (hunk 1 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler_final_check.cpp:451]: I added some comments. Please check it.

Comment (Adda0) [reply src/smt/theory_str_noodler/theory_str_noodler_final_check.cpp:451]: Looks great. I am happy with the changes.

Comment (jurajsic) [reply src/smt/theory_str_noodler/theory_str_noodler_final_check.cpp:451]: Looks good.

**Comment by vhavlena** [comment]
Can I merge it?

**Pull request merged**